AEPCore/Tests/FunctionalTests/RulesEngineFunctionalTests.swift

```diff
@@ -871,6 +872,367 @@ class RulesEngineFunctionalTests: XCTestCase {
+        let dispatchedEvent1 = mockRuntime.dispatchedEvents[0]
+        mockRuntime.dispatchedEvents.removeAll()
+
+        // Process launch event
+        _ = rulesEngine.process(event: eventLaunch)
+        XCTAssertEqual(1, mockRuntime.dispatchedEvents.count)
+        let dispatchedEvent2 = mockRuntime.dispatchedEvents[0]
+        mockRuntime.dispatchedEvents.removeAll()
+
+        // Process first dispatched event; dispatch chain count = 1
+        // Expect dispatch to fail as max allowed chained events is 1
+        _ = rulesEngine.process(event: dispatchedEvent1)
+        XCTAssertEqual(0, mockRuntime.dispatchedEvents.count)
+
+        // // Process second dispatched event; dispatch chain count = 1
+        // Expect dispatch to fail as max allowed chained events is 1
+        _ = rulesEngine.process(event: dispatchedEvent2)
+        XCTAssertEqual(0, mockRuntime.dispatchedEvents.count)
+    }
+
+    func testDispatchEvent_processedEventMatchesMultipleDispatchConsequences() {
+        /// Given: two launch rules with the same consequence but different conditions
+
+        //    ---------- dispatch event rule 1 condition ----------
+        //        "conditions": [
+        //        {
+        //          "type": "matcher",
+        //          "definition": {
+        //            "key": "~type",
+        //            "matcher": "eq",
+        //            "values": [
+        //              "com.adobe.eventType.edge"
+        //            ]
+        //          }
+        //        },
+        //        {
+        //          "type": "matcher",
+        //          "definition": {
+        //            "key": "~source",
+        //            "matcher": "eq",
+        //            "values": [
+        //              "com.adobe.eventSource.requestContent"
+        //            ]
+        //          }
+        //        }
+        //      ]
+        //    ---------- dispatch event rule 1 consequence ----------
+        //        "detail": {
+        //           "type" : "com.adobe.eventType.edge",
+        //           "source" : "com.adobe.eventSource.requestContent",
+        //           "eventdataaction" : "copy"
+        //         }
+        //    --------------------------------------
+
+        //    ---------- dispatch event rule 2 condition ----------
+        //        "conditions": [
+        //          {
+        //            "type": "matcher",
+        //            "definition": {
+        //              "key": "dispatch",
+        //              "matcher": "eq",
+        //              "values": [
+        //                "yes"
+        //              ]
+        //            }
+        //          }
+        //        ]
+        //    ---------- dispatch event rule 2 consequence ----------
+        //        "detail": {
+        //           "type" : "com.adobe.eventType.edge",
+        //           "source" : "com.adobe.eventSource.requestContent",
```

Comment:
❓ do we want to keep these Edge/Lifecycle test cases in AEPCore RulesEngineFunctionalTests or should we make these tests more generic instead and cover these as integration tests

Reply:
I like having these functional tests as they run fast and it's easier to run negative tests than with integration tests. I've just added a few integration tests as a sanity check to verify all the end-to-end plumbing is hooked up.